### PR TITLE
Getting the docs up and running

### PR DESCRIPTION
**Pull request by pkofod**
It's easier to comment on the docs inline if there's a PR, as mentioned in #224 .

### CODE REVIEW

File: docs/src/algo/gradientdescent.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+Gradient Descent a common name for a quasi-Newton solver. This means that it takes
+steps according to
+
+$ x_{n+1} = x_n - P^{-1}\nabla f(x_n)$
+
+where $P$ is the Hessian in Newton's method, but is generally just a positive definite
+matrix. In Gradient Descent, $P$ is simply an appropriately dimensioned identity matrix.
+This means that we go in the exact opposite direction of the gradient. This means
+that we do not use the curvature information from the Hessian, or an approximation
+of it. While it does seem quite logical to go in the opposite direction of the fastest
```
Comment (KristofferC) [inline docs/src/algo/gradientdescent.md]: Repetition of "This means" in two consecutive sentences.

File: docs/src/algo/newton.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+The constructor takes one keyword
+
+* `linesearch! = a(d, x, p, x_new, g_new, lsr, c, mayterminate)`, a function performing line search, see the line search section.
+
+## Description
+Newton's method for optimization has a long history, and is in some sense the
+gold standard in unconstrained optimization of smooth functions, at least from a theoretical viewpoint.
+The main benefit is that it has a quadratic rate of convergence near a local optimum. The main
+cost is that the user has to provide a Hessian. This can be difficult, complicated, or simply annoying.
+Tools such as automatic differentiation can help, and it obviously requires that the function is
```
Comment (KristofferC) [inline docs/src/algo/newton.md]: This sentence feels weird, automatic differentiation and continuity requirement are not related.

File: docs/src/algo/newton.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+where the inverse is not calculated directly, but the step size is instead calculated by solving
+
+$ H(x) \textbf{s} = \nabla f(x_n) $.
+
+This is equivalent to minimizing a quadratic model, $m_k$ around the current $x_n$
+
+$ m_k(s) = f(x_n) + \nabla f(x_n)^\top \textbf{s} + \frac{1}{2} \textbf{s}^\top H(x_n) \textbf{s} $
+
+For functions where $H(x_n)$ is difficult, or computationally expensive, to optain, we might
```
Comment (KristofferC) [inline docs/src/algo/newton.md]: -> "or computationally expensive to obtain"

File: docs/src/algo/newton.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+replace the Hessian with another positive definite matrix that approximates it.
+Such methods are called Quasi-Newton methods; see (L-)BFGS and Gradient Descent.
+
+In a sufficiently small neighborhood around the minimizer, Newton's method has
+quadratic convergence, but globally it might have slower convergence,or it might
```
Comment (KristofferC) [inline docs/src/algo/newton.md]: space after `,`

File: docs/src/algo/newton.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+even diverge. Make ensure convergence, a line search is performed for each $\textbf{s}$.
```
Comment (KristofferC) [inline docs/src/algo/newton.md]: "Make ensure convergence"?

File: docs/src/algo/gradientdescent.md [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+                  P = nothing,
+                  precondprep! = (P, x) -> nothing)
+```
+## Description
+Gradient Descent a common name for a quasi-Newton solver. This means that it takes
+steps according to
+
+$ x_{n+1} = x_n - P^{-1}\nabla f(x_n)$
+
+where $P$ is the Hessian in Newton's method, but is generally just a positive definite
```
Comment (KristofferC) [inline docs/src/algo/gradientdescent.md]: "where $P$ *typically* is the"?

File: docs/src/algo/lbfgs.md [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+```julia
+LBFGS(; m = 10,
+        linesearch! = hz_linesearch!,
+        P = nothing,
+        precondprep! = (P, x) -> nothing)
+```
+## Description
+This means that it takes steps according to
+
+$ x_{n+1} = x_n - P^{-1}\nabla f(x_n)$
```
Comment (KristofferC) [inline docs/src/algo/lbfgs.md]: This is not right?

File: docs/src/algo/precondition.md [hunk 8]
```diff
@@ -0,0 +1,58 @@
+# Preconditioning
+
+The `GradientDescent`, `ConjugateGradient` and `LBFGS` methods support preconditioning. A preconditioner
+can be thought of as a change of coordinates under which the Hessian is better conditioned. With a
+"good" preconditioner substantially improved convergence is possible.
```
Comment (KristofferC) [inline docs/src/algo/precondition.md]: why quotations marks around "good"?

File: docs/src/algo/precondition.md [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+* `A_ldiv_B!(pgr, P, gr)` : apply `P` to a vector `gr` and store in `pgr`
+      (intuitively, `pgr = P \ gr`)
+* `dot(x, P, y)` : the inner product induced by `P`
+      (intuitively, `dot(x, P * y)`)
+
+Precisely what these operations mean, depends on how `P` is stored. Commonly, we store a matrix `P` which
+approximates the Hessian in some vague sense. In this case,
+
+* `A_ldiv_B!(pgr, P, gr) = copy!(pgr, P \ A)`
```
Comment (KristofferC) [inline docs/src/algo/precondition.md]: Ref https://github.com/JuliaLang/julia/issues/11325. When this is done, the `copy!` part should not be needed.

File: docs/src/algo/precondition.md [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+* `dot(x, P, y) = dot(x, P * y)`
+
+Finally, it is possible to update the preconditioner as the state variable `x`
+changes. This is done through  `precondprep!` which is passed to the
+optimisers as kw-argument, e.g.,
```
Comment (KristofferC) [inline docs/src/algo/precondition.md]: optimizers

File: docs/src/index.md [hunk 11]
```diff
@@ -0,0 +1,40 @@
+# Optim.jl
+## How
+Optim is registered in [METADATA.jl](https://github.com/JuliaLang/METADATA.jl).
+This means that all you need to do to install Optim, is to run
+```julia
+Pkg.add("Optim")
+```
+## What
+Optim is a package, written entirely in Julia, for optimizing functions of
```
Comment (KristofferC) [inline docs/src/index.md]: "Optim is a *Julia* package"? sounds a bit strange to just call it a package.

File: docs/src/index.md [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+various kinds. While there is some support for box constrained optimization, most
+of the solvers tries to find an $x$ that minimizes a function $f(x)$ without any constraints.
+ That is the main focus is on unconstrained optimization.
```
Comment (KristofferC) [inline docs/src/index.md]: "Thus, the main focus of this package is..." maybe?

File: docs/src/index.md [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+
+## Why
+There are many solvers available from both free and commercial sources, and many
+of them are accessible from Julia. Few of them are written in Julia.
+Performance-wise this is rarely a problem, as they are often written in either
+Fortran or C. However, solvers written directly in Julia
+does come with some advantages when Julia is the main language used.
```
Comment (KristofferC) [inline docs/src/index.md]: It feels like you are saying something like:

> Solvers written in Julia comes with advantage when they are written in Julia

Maybe you can take away the stuff after "advantages".

File: docs/src/index.md [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+When writing Julia software (packages) that require something to be optimized, the programmer
+can either choose to write their own optimization routine, or use one of the many
+available solvers. For example, this could be something from the [NLOpt](...) suite.
+This means adding a dependency which is not written in Julia, and more assumptions
+have to be made as to the environment the user is in. Does the user have the proper
+compilers? Is it possible to use GPL'ed code in the project? Optim is released
+under the MIT license, and installation is a simple `Pkg.add`, so it really doesn't
+get much freer, easier, and lightweight than that.
+
+It is also true, that using a solver written in C or Fortran makes it impossible to leverage one
```
Comment (KristofferC) [inline docs/src/index.md]: I think automatic differentiation would be useful to mention in this section.

File: docs/src/user/config.md [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+
+Special methods for univariate optimization:
+
+* `Brent()`
+* `GoldenSection()`
+
+### General Options
+In addition to the solver, you can alter the behavior of the Optim package by using the following keywords:
+
+* `x_tol`: What is the threshold for determining convergence? Defaults to `1e-32`.
```
Comment (KristofferC) [inline docs/src/user/config.md]: convergence in what?

File: docs/src/user/minimization.md [hunk 16]
```diff
@@ -0,0 +1,126 @@
+## Minimizing a function
+To show how the Optim package can be used, we'll implement the
+[Rosenbrock function](http://en.wikipedia.org/wiki/Rosenbrock_function),
+a classic problem in numerical optimization. We'll assume that you've already
```
Comment (KristofferC) [inline docs/src/user/minimization.md]: "We'll" -> "We" since the assumption is now?

File: docs/src/user/minimization.md [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+installed the Optim package using Julia's package manager.
+First, we'll load Optim and define the Rosenbrock function:
```
Comment (KristofferC) [inline docs/src/user/minimization.md]: "we'll" -> "we"?

File: docs/src/user/results.md [hunk 18]
```diff
@@ -0,0 +1,48 @@
+## Obtaining results
```
Comment (KristofferC) [inline docs/src/user/results.md]: This whole file seems redundant when looking at `minimization.md`

File: docs/src/user/tipsandtricks.md [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+        pred_i = betas[1] + betas[2] * X[i]
+        err += (Y[i] - pred_i)^2
+    end
+    return err
+end
+```
+and then optimize the following anonymous function
+```jl
+res = optimize(b -> sqerror(b, x, y), [0.0, 0.0])
+```
```
Comment (KristofferC) [inline docs/src/user/tipsandtricks.md]: Maybe mention that this is typically called a *closure* to introduce the terminology?

File: docs/src/algo/newton.md (hunk 2 above)
Comment (pkofod) [inline docs/src/algo/newton.md]: I agree, it looks very weird... I'll figure out what I meant, and rewrite it.

File: docs/src/algo/gradientdescent.md (hunk 6 above)
Comment (pkofod) [inline docs/src/algo/gradientdescent.md]: I'll reword it.

File: docs/src/algo/precondition.md (hunk 8 above)
Comment (pkofod) [inline docs/src/algo/precondition.md]: not sure, this was lifted from the readme.

File: docs/src/algo/precondition.md (hunk 9 above)
Comment (pkofod) [inline docs/src/algo/precondition.md]: Looking forward to it!

File: docs/src/algo/precondition.md (hunk 10 above)
Comment (pkofod) [inline docs/src/algo/precondition.md]: I think Christoph is in Britain, but I'll change it :)

File: docs/src/algo/lbfgs.md (hunk 7 above)
Comment (pkofod) [inline docs/src/algo/lbfgs.md]: Why? I mean, sure it gets calculated in a smart way, but why isn't that "what's being done"?

**Comment by pkofod** [comment]
I think Documenter might have been updated :) It looks different now, but in a good way.

**Comment by pkofod** [comment]
Commented out some sections I hadn't written yet. I need to be a bit more sure what is going on in conjugate gradient to formulate something sensible. I know how it works, but I'm not sure how to present it in a paragraph or two.

**Comment by pkofod** [comment]
I'm going to push a version of this asap, and then the missing solvers can be added over the summer. I am just trying to find out how to get the travis-things for Documenter set without admin rights :)

**Comment by pkofod** [comment]
Let's see if it works (and if it doesn't, hope that Hatherly is not too busy with JCon :))

**Pull request merged**